pythia/models/butd.py

```diff
@@ -108,52 +108,87 @@ def get_data_t(self, t, data, batch_size_t, prev_output):
 
         return data, batch_size_t
 
-    def forward(self, sample_list):
-        # Stores the output probabilites. Not used if beam_search inference
-        scores = sample_list.answers.new_ones(
-            (
-                sample_list.answers.size(0),
-                self.text_processor.max_length,
-                self.vocab_size,
-            ),
-            dtype=torch.float,
-        )
+    def forward_greedy(self, sample_list, scores):
 
-        # For beam search inference. Currently beam seach for BUTD works only 
-        # with batch_size = 1 and should be used with run_type inference only.
-        # TODO : Implement batch beam search
-        if self.config["inference"]["type"] == "beam_search":
-            beam_search = BeamSearch(
-                self.vocab, self.config["inference"]["params"]["beam_length"]
+        batch_size = sample_list.image_feature_0.size(0)
+        data, sample_list, timesteps = self.prepare_data(sample_list, batch_size)
+        output = None
+        batch_size_t = batch_size
+        for t in range(timesteps):
+            data, batch_size_t = self.get_data_t(t, data, batch_size_t, output)
+            pi_t = data["texts"][:, t].unsqueeze(-1)
+            embedding = self.word_embedding(pi_t)
+            attention_feature, _ = self.process_feature_embedding(
+                "image", sample_list, embedding[:, 0, :], batch_size_t=batch_size_t
             )
-            sample_list = beam_search.init_batch(sample_list)
+            output = self.classifier(attention_feature)
+            scores[:batch_size_t, t] = output
 
+        model_output = {"scores": scores}
+        return model_output
+
+    def forward_beam_search(self, sample_list, scores):
+        beam_search = BeamSearch(
```

Comment:
Now afterwards this will be initialized using constructor pattern, like how we construct the processors.

Reply:
@apsdehal @dexter1691 I've created a common forward but had to use a conditional statement because the calls to create a decoder object had different parameter requirements. If the differences were only limited to object class name then conditional statements could have been avoided.